crates/sui-indexer/src/store/pg_indexer_store.rs

```diff
@@ -1295,6 +1328,84 @@ impl PgIndexerStore {
         Ok(())
     }
 
+    async fn update_watermarks(&self, watermarks: Vec<Watermark>) -> Result<(), IndexerError> {
+        use diesel_async::RunQueryDsl;
+
+        let guard = self
+            .metrics
+            .checkpoint_db_commit_latency_watermarks
+            .start_timer();
+
+        let updates_count = watermarks.len();
+
+        // UpperBound, LowerBound, and Tail updates require different `ON CONFLICT DO UPDATE`
+        // clauses, so split them into their own vectors and process each in bulk. TODO: two
+        // optimizations
+        // 1. multiple updates of the same type for one `entity` can be reduced to the latest one
+        // 2. multiple updates of different kinds for one `entity` can be merged into a single
+        //    update
+        let mut map: HashMap<WatermarkUpdateType, Vec<StoredWatermark>> = HashMap::new();
+        for watermark in watermarks {
+            map.entry(watermark.update_type())
+                .or_default()
+                .push(StoredWatermark::from(watermark));
+        }
+
+        transaction_with_retry(&self.pool, PG_DB_COMMIT_SLEEP_DURATION, |conn| {
+            async {
+                if let Some(upper_bound_updates) = map.get(&WatermarkUpdateType::UpperBound) {
+                    diesel::insert_into(watermarks::table)
+                        .values(upper_bound_updates)
+                        .on_conflict(watermarks::entity)
+                        .do_update()
+                        .set((
+                            watermarks::hi.eq(excluded(watermarks::hi)),
+                            watermarks::epoch_hi.eq(excluded(watermarks::epoch_hi)),
+                        ))
+                        .execute(conn)
+                        .await?;
+                }
+
+                if let Some(lower_bound_updates) = map.get(&WatermarkUpdateType::LowerBound) {
+                    diesel::insert_into(watermarks::table)
+                        .values(lower_bound_updates)
+                        .on_conflict(watermarks::entity)
+                        .do_update()
+                        .set((
+                            watermarks::lo.eq(excluded(watermarks::lo)),
+                            watermarks::epoch_lo.eq(excluded(watermarks::epoch_lo)),
+                            watermarks::timestamp_ms.eq(excluded(watermarks::timestamp_ms)),
```

Comment:
I think we need to be a bit more careful about when we set the lowerbounds, because we can't lower the lowerbound or the timestamp after we have raised it.